File: elipseex.py
import numpy as np
import matplotlib.pyplot as plt

def plot_ellipse(a, b, excentricidade):
    """
    Plota uma elipse com base em seus semi-eixos e excentricidade.

    Args:
        a (float): O semi-eixo maior.
        b (float): O semi-eixo menor.
        excentricidade (float): A excentricidade da elipse (e).
    """

    # Cria uma matriz de ângulos de 0 a 2*pi
    theta = np.linspace(0, 2 * np.pi, 100)

    # Coordenadas da elipse usando as equações paramétricas
    x = a * np.cos(theta)
    y = b * np.sin(theta)

    # Cria a figura e o eixo do plot
    fig, ax = plt.subplots(figsize=(8, 8))

    # Plota a elipse
    ax.plot(x, y, label=f'Excentricidade = {excentricidade:.2f}')

    # Adiciona a legenda
    ax.legend()

    # Define o título do gráfico
    ax.set_title(f'Gráfico de uma Elipse com Excentricidade Alta')

    # Define os rótulos dos eixos
    ax.set_xlabel('Eixo X')
    ax.set_ylabel('Eixo Y')

    # Define a proporção dos eixos para que a elipse não pareça distorcida
    ax.set_aspect('equal', adjustable='box')

    # Adiciona as linhas de grade
    ax.grid(True, linestyle='--')

    # Mostra o gráfico
    plt.show()

# --- Exemplo de uso ---
# Elipse com excentricidade alta (próxima de 1)
# O semi-eixo maior (a) é 5 e o semi-eixo menor (b) é 1.
# Calculamos a excentricidade (e) com a fórmula: e = sqrt(1 - (b^2/a^2))
a = 5.0
b = 1.0
excentricidade = np.sqrt(1 - (b**2 / a**2))

# Chama a função para plotar a elipse
plot_ellipse(a, b, excentricidade)

File: elipseex.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define NUM_PONTOS 1000
#define PI 3.14159265358979323846

int main() {
    // Definindo os semi-eixos para uma elipse com alta excentricidade
    // Excentricidade (e) = sqrt(1 - (b^2 / a^2))
    // Quando 'b' é muito menor que 'a', 'e' se aproxima de 1.
    double a = 5.0; // Semi-eixo maior
    double b = 1.0; // Semi-eixo menor

    // Calculando a excentricidade
    double excentricidade = sqrt(1.0 - (b * b) / (a * a));

    // Abrindo um arquivo para salvar os pontos da elipse
    FILE *fp = fopen("dados_elipse.txt", "w");
    if (fp == NULL) {
        perror("Erro ao abrir o arquivo");
        return 1;
    }

    // Gerando os pontos da elipse
    for (int i = 0; i <= NUM_PONTOS; i++) {
        double theta = (double)i / NUM_PONTOS * 2.0 * PI;
        double x = a * cos(theta);
        double y = b * sin(theta);
        fprintf(fp, "%.4f %.4f\n", x, y);
    }

    // Fechando o arquivo
    fclose(fp);

    printf("Pontos da elipse salvos em dados_elipse.txt\n");
    printf("Gerando o gráfico com Gnuplot...\n");

    // Abrindo um pipe para o Gnuplot e enviando os comandos
    FILE *gnuplotPipe = popen("gnuplot -persistent", "w");
    if (gnuplotPipe == NULL) {
        perror("Erro ao abrir pipe para o Gnuplot");
        return 1;
    }

    // Enviando os comandos de plotagem para o Gnuplot
    fprintf(gnuplotPipe, "set title 'Gráfico de uma Elipse com Excentricidade Alta (e = %.2f)'\n", excentricidade);
    fprintf(gnuplotPipe, "set xlabel 'Eixo X'\n");
    fprintf(gnuplotPipe, "set ylabel 'Eixo Y'\n");
    fprintf(gnuplotPipe, "set size ratio -1\n"); // Força a proporção 1:1 para evitar distorção
    fprintf(gnuplotPipe, "plot 'dados_elipse.txt' with lines notitle\n");
    
    // Fechando o pipe
    pclose(gnuplotPipe);

    return 0;
}